fix hypot with more than two arguments

cossio opened the pull request:
Fixes  #27141: The previous code led to under/overflow.

Before this change:

```
a = 10^10
hypot(a, a, a, a)
# ERROR: DomainError with -5.828369621610136e18:
```

After this change:

```
a = 10^10
hypot(a,a,a,a)
# 2.0e10
```

Commit:
fix hypot with multiple arguments

simonbyrne commented:
Can you add a test case?

There probably are more efficient ways to do this (similar to how `norm` is implemented in LinearAlgebra), but this is a net improvement.

cossio commented:
@simonbyrne Added a test.

You are right this is correct but about 5 times slower than the previous implementation. I can try to make it faster with some advice.

On the other hand, is there a reason against `hypot(x...) = norm(x)`?

cossio commented:
It seems to me then that `hypot` could be removed in favor of `norm`.

stevengj commented:
See also the discussions in #27141 and #27251.

I would have thought that something like this would be faster
```jl
function hypot2(x::Number...)
    maxabs = float(maximum(abs, x))
    (iszero(maxabs) || isinf(maxabs)) && return maxabs
    return maxabs * sqrt(sum(y -> abs2(y/maxabs), x))
end
```
but `@btime` says it is allocating for some reason when I call `hypot2(3,4,5)`, even though `@code_warntype` doesn't find any type instabilities.  What's going wrong?

stevengj commented on base/math.jl:
```diff
@@ -567,7 +567,7 @@ end
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = reduce(hypot, x)
```
This will give the wrong results for `hypot(-2)`.   It may be simplest to define the special case `hypot(x::Number) = abs(x)`.

cossio replied:
Added `hypot(x::Number) = abs(x)`.

cossio replied:
Although I'm not 100% sure if `hypot` with a single argument makes a lot of sense.

cossio commented:
Is there any difference between `hypot(x...)` and `norm(x)`? What's the point of having two functions here?

cossio commented:
@stevengj `maximum` allocates. See #28928

simonbyrne commented:
> Is there any difference between `hypot(x...)` and `norm(x)`? What's the point of having two functions here?

`norm` is the reducing version of `hypot` (similar to `maximum` vs `max, `sum` vs `+`, etc). `norm` also includes much more functionality (e.g. handling arrays of arrays).

simonbyrne commented:
One way forward would be to write an efficient implementation of `mapreduce(f, ::typeof(hypot), x)` in Base, then define `hypot(x...) = mapreduce(identity, hypot, x)` and `LinearAlgebra.generic_norm2(x) = mapreduce(norm, hypot, x)`.

andreasnoack commented:
As a side note here, I was recently reading up on MKL's super fast reduction to bi- and tridiagonal matrices. They mentioned that, at some point, one of the bottlenecks become `norm` and for MKL they have implemented a different `norm` that speculatively computes `nrmsq = dot(x,x)` and checks if the value is "safe" and then only computes the slow scaled version if `nrmsq` is outside the safe bounds and otherwise just returns `sqrt(nrmsq)`. In most cases, this is much faster and in rare cases, it is much slower.

stevengj commented:
@cossio, `maximum` only allocates when the result is an array, not when the result is a scalar.
```jl
julia> x = rand(10);

julia> @btime maximum($x);
  15.835 ns (0 allocations: 0 bytes)
```
That's why it confuses me that it is allocating here.

cossio commented:
@stevengj Could be because of the generator in the sum.

```
function hypot3(x::Number...)
           maxabs = float(maximum(abs, x))
           (iszero(maxabs) || isinf(maxabs)) && return maxabs
           t = zero(maxabs)
           for y in x
               t += abs2(y/maxabs)
           end
           return maxabs * t
end
```

allocates less:

```
julia> @btime hypot2(3,4,5)
  2.630 μs (14 allocations: 384 bytes)
7.0710678118654755

julia> @btime hypot3(3,4,5)
  1.917 μs (8 allocations: 272 bytes)
10.0
```

cossio commented:
And this version is still better:

```julia
function hypot4(x::Number...)
           maxabs = -Inf
           for y in x
               maxabs = max(maxabs, abs(y))
           end
           (iszero(maxabs) || isinf(maxabs)) && return maxabs
           t = zero(maxabs)
           for y in x
               t += abs2(y/maxabs)
           end
           return maxabs * sqrt(t)
end
```

```
julia> @btime hypot4(3,4,5)
  28.573 ns (1 allocation: 16 bytes)
7.0710678118654755
```

stevengj commented:
This is the fastest implementation I've found so far, based on @thchr's comment about splatting and @andreasnoack's comment about speculative execution:
```
function _hypot6(x::NTuple{N,<:Number}) where N
    simple = sum(y -> abs2(float(y)), x)
    isinf(simple) || iszero(simple) || return sqrt(simple)
    maxabs = float(maximum(abs, x))
    (iszero(maxabs) || isinf(maxabs)) && return maxabs
    return maxabs * sqrt(sum(y -> abs2(y/maxabs), x))
end
hypot6(x::Number...) = _hypot6(promote(x...))
```
It's about 30% faster than `reduce(hypot, x)` for 3 arguments and about 3× faster for 5 arguments.

However, type inference is failing for it: @code_warntype says that the inferred return type for `_hypot6((3,4,5))` is `Any`, which is mysterious to me.  Is this an inference bug?

thchr commented:
Building again on @stevengj's latest version, this does not allocate and infers - and is faster still - but, admittedly, is not a very pretty thing: 
```julia
function _hypot7(x::NTuple{N,<:Number}) where N
    simple = sum(y -> abs2(float(y)), x)
    isinf(simple) || iszero(simple) || return sqrt(simple)
    return __hypot7(x)
end
function __hypot7(x::NTuple{N,<:Number}) where N
    maxabs = float(maximum(abs, x))
    (iszero(maxabs) || isinf(maxabs)) && return maxabs
    return maxabs * sqrt(sum(y -> abs2(y/maxabs), x))
end
hypot7(x::Number...) = _hypot7(promote(x...))
```
```julia
@btime hypot7(3,4,5)
  10.899 ns (0 allocations: 0 bytes)
7.0710678118654755
```

cossio commented:
@thchr very nice! But I don't see why this is slower than the original version,

```julia
hypot0(x::Number...) = sqrt(sum(abs2(y) for y in x))
julia> @btime hypot0(3,4,5)
  3.990 ns (0 allocations: 0 bytes)
7.0710678118654755

julia> @btime hypot7(3,4,5)
  11.216 ns (0 allocations: 0 bytes)
7.0710678118654755
```

thchr commented:
> > @cossio, there is the overhead of the extra `isinf` and `iszero` checks.
> 
> Those seem negligible. See edit in my comment just above yours.

That particular benchmark is very specific to having `::Integer` inputs, however. The `simple = sum(y -> abs2(float(y)), x)` is actually somewhat faster than the generator version for `::Float64` inputs (I guess float-input would be more common; though, overall, I guess the vararg `hypot` function is unlikely to see a lot of action). Still, the difference is a little odd...

Quite a few unexpected inference gotcha's and fragility in this example, somehow...

cossio commented:
> The ` sum(y -> abs2(float(y)), x)` and `sum(abs2(y) for y in x)` differences for `::Float64` arguments are probably just noise. The former is as fast as the latter also for `::Int` arguments if the `float` conversion is dropped; I guess it actually isn't needed here, or?

Then we can just leave `sum(abs2(y) for y in x)`? Looks more readable to me.

stevengj commented:
> The former is as fast as the latter also for `::Int` arguments if the `float` conversion is dropped; I guess it actually isn't needed here, or?

The `float` conversion is needed *before* computing `abs2`, because otherwise you can get spurious wraparound:

```
julia> abs2(10^18)
-5527149226598858752

julia> abs2(float(10^18))
1.0e36
```
Indeed, there should probably be a `hypot` test for `hypot(i,i)` and `hypot(i,i,i)` where `i = typemax(Int)`.

cossio commented:
@stevengj Ok. Also added the test for `typemax(Int)`.

cossio commented on base/math.jl:
```diff
@@ -567,7 +567,18 @@ end
+# https://github.com/JuliaLang/julia/pull/30301#issuecomment-445944572
+function _hypot(x::NTuple{N,<:Number}) where N
+    simple = sum(abs2(float(y)) for y in x)
+    isinf(simple) || iszero(simple) || return sqrt(simple)
```
I followed your suggestion. But it still doesn't work for `realmax`:

```julia
julia> hypot(floatmax(), floatmax())
Inf
```

(this is calling the last committed version)

stevengj replied:
@cossio, `Inf` is the correct answer for `hypot(floatmax(), floatmax())`.

Computing `hypot(floatmax(), floatmax())` in exact arithmetic would give `floatmax()*√2 > floatmax()`, so the result has no choice but to overflow to `Inf`.

stevengj replied:
(Note that this requires #30393 to give correct results for `BigFloat`.)

stevengj replied:
It should be `<=` … sorry, that occurred to me last night but I forgot to ping you about it.

cossio replied:
@stevengj Any reason for `<` instead of `<=`?

stevengj replied:
No, please change it to `<=`

cossio commented:
By the way, I'm leaving the two argument method because for some reason it seems 4x faster than the generic version.

cossio commented on test/math.jl:
```diff
@@ -249,10 +250,12 @@ end
             @test hypot(T(0), T(0)) === T(0)
             @test hypot(T(Inf), T(Inf)) === T(Inf)
             @test hypot(T(Inf), T(x)) === T(Inf)
-            @test hypot(T(Inf), T(NaN)) === T(Inf)
```
@test hypot(T(Inf), T(NaN)) === T(Inf) looked strange. Anyone knows why this test is here? Shouldn't this return NaN?

simonbyrne replied:
From the C-2011 spec:
> hypot(±∞, y) returns +∞, even if y is a NaN.

simonbyrne replied:
And the IEEE754-2008 spec:
> For the hypot function, hypot(±0, ±0) is +0, hypot(±∞, qNaN) is +∞, and hypot(qNaN, ±∞) is +∞.

andreasnoack replied:
I actually don't remember adding these test but as @simonbyrne just pointed out on Slack, `Inf` is following the standard. The reasoning is that `NaN` is used when it cannot be determined if the limit is `0` or `+-Inf`. In either case, the result of `hypot` is `Inf` when the other input is `+-Inf` so returning `Inf` here makes sense.

stevengj commented on test/math.jl:
```diff
@@ -249,10 +249,12 @@ end
             @test hypot(T(Inf), T(Inf)) === T(Inf)
             @test hypot(T(Inf), T(x)) === T(Inf)
             @test hypot(T(Inf), T(NaN)) === T(Inf)
+            @test hypot(T(1e10), T(1e10), T(1e10), T(1e10)) ≈ 2e10
```
for example, I would add a test like:
```jl
for scale in (zero(T), floatmin(T)*1e3, floatmax(T)*1e-3, T(Inf))
    @test hypot(1*scale,2*scale) ≈ hypot(1,2)*scale rtol=eps(T)*8
    @test hypot(1*scale,2*scale,3*scale) ≈ hypot(1,2,3)*scale rtol=eps(T)*8
end
```

cossio replied:
Done

cossio commented:
I made some changes so that `hypot(x...)` returns `Inf` if any arg is `Inf`, even in presence of `NaN`s.

cossio commented:
I'm getting failures from tests of involving `norm`. Does `norm` depend on `hypot`?

stevengj commented on base/math.jl:
```diff
@@ -568,11 +567,18 @@ end
+	else
+		return sqrt(simple)
+	end
 end
```
Probably faster to do something like:
```jl
simple = sum(abs2(float(y)) for y in x)
isinf(simple) || simple ≤ floatmin(simple) * (N-1) || return sqrt(simple)
isnan(simple) && return any(isinf, y) ? oftype(y, Inf) : y
return __hypot(x)
```
This way you only add one additional check in the common case where no NaN values are encountered.

cossio commented:
@simonbyrne @andreasnoack @thchr @stevengj I did some more changes. I think it's almost as fast as the previous version (whenever the speculative naive approach works).

thchr commented:
> > @cossio, `maximum` only allocates when the result is an array, not when the result is a scalar.
> 
> It still allocates if the container's `eltype` is heterogeneous:

That should not have any impact here though, since we `promote(x...)` the input tuple.

cossio commented on base/math.jl:
```diff
@@ -536,38 +536,48 @@ Stacktrace:
+	r = sqrt(s)
+	if isnan(s) || isinf(s) || s ≤ floatmin(s) * (N-1)
+		# if overflow/underflow, try normalization
+		any(isinf, x) && return oftype(r, Inf) # IEEE 754
```
@stevengj I think this is what you are suggesting. I also tried `isnan(s) && any(isinf, x) && return oftype(r, Inf)`, but that fails this test: `@test hypot(1s, 2s) ≈ s * hypot(1,2)   rtol=8eps(T)` with `T=Float32`.

stevengj replied:
Shouldn't that be
```
if isnan(s)
    return any(isinf, x) ? oftype(r, Inf) : r
end
```
Seems like this is better for an underflow case like `hypot(0,0,0,0,0,0,0)` so that it avoids checking `isinf` on all the arguments.

stevengj replied:
(Seems better just to write out `sqrt(s)` in several places rather than run the risk of computing it unnecessarily.)

cossio replied:
Done. But I think we compute `r = sqrt(s)` in all execution paths. Unless `oftype(f(x), y)` is able to elide evaluation of `f(x)` and just infer its return type?

thchr commented on base/math.jl:
```diff
@@ -536,38 +536,48 @@ Stacktrace:
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number, y::Number...) = _hypot(promote(x, y...))
```
What's the motivation for this? If this is to disambiguate the 1-arg method (is that really needed?), then it doesn't seem worthwhile: calling `hypot` with a single argument should be an extremely rare occurrence; imo, it doesn't seem worth it to complicate this method further for such a rare thing (even if a slightly faster method exists)

stevengj replied:
It's not unreasonable to me just to define `hypot() = 0`.

cossio replied:
@stevengj Having `hypot()` seems inconsistent with other functions such as `max`, `min`, which err with no arguments. The previous version of `hypot` (before this PR) also errors without arguments. Also, why `0::Int` is the correct answer? Why not `0.0`, `false`, `Int16(0)`?

I think IEEE doesn't help here since it only specifies the two argument version?

thchr replied:
Alternatively, the `hypot()` call corresponds to an `N = 0` case; since `N` is a known type parameter, that case could be inferred and treated separately statically: i.e., it shouldn't cost anything to test `N == 0 && error( ... )` inside the function - the test would be removed at compilation for `N != 0`, AFAIK.

stevengj replied:
I'm less worried about `0::Int` (since numeric zeros interoperate) than someone extending `hypot` for non-`Number` types, or using it for dimensionful types, in which case `0` is not an additive identity.

cossio replied:
@thchr I removed the one arg version and did the `N==0` error case as you suggested

stevengj commented on base/math.jl:
```diff
@@ -536,59 +536,46 @@ Stacktrace:
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	r = sqrt(s)
+	if isnan(s) || isinf(s) || s ≤ floatmin(s) * (N-1)
```
`isnan(s) || isinf(s)` is equivalent to `!isfinite(s)`

stevengj replied:
Or, to avoid the negation:
```jl
isfinite(s) && s > floatmin(s) * (N-1) && return sqrt(s)
```

stevengj commented:
Given the discussion in #30393, it may be that we can't use `floatmin` here, and should instead replace `floatmin(s)` with `_floatmin(s)` where we define:
```jl
_floatmin(s::IEEEFloat) = floatmin(s)
_floatmin(s::AbstractFloat) = nextfloat(zero(s)) / eps(typeof(s))
```
To make sure it works correctly for `BigFloat` and other types in underflow circumstances, there should be a test like:
```jl
for T in (Float16, Float32, Float64, BigFloat)
    x = sqrt(floatmin(T))*2
    @test hypot(x,x/4) ≈ x * sqrt(17/T(16))
    @test hypot(x,x/4,x/4) ≈ x * sqrt(9/T(8))
end
```
in our loop over floating

cossio commented:
@stevengj Your version doesn't pass, even with the modified `_floatmin`. Shouldn't the test be:

```
@testset "$T" for T in (Float16, Float32, Float64, BigFloat)
    x = sqrt(floatmin(T))*1.1
    @test hypot(x,x/4) ≈ x * sqrt(BigFloat(17//16))
    @test hypot(x,x/4,x/4) ≈ x * sqrt(BigFloat(9//8))
end
```

But this passes also with the current version of `floatmin` in Julia.

stevengj commented:
`sqrt(17/BigFloat(16))` and `sqrt(BigFloat(17//16))` give the same answer, so I'm not sure what your change accomplishes?   The tests (neither my version nor your version) fail on Julia 1.0 for me.

cossio commented:
@stevengj  This:

```
@test hypot(x,x/4) ≈ x * sqrt(17/T(16))
@test hypot(x,x/4,x/4) ≈ x * sqrt(9/T(8))
```

fails when `T` is `Float16`. The only change I did with respect to your version was that the right-hand side was always evaluated in high-precision.

stevengj commented:
Maybe a better test:
```
for T in (Float16, Float32, Float64, BigFloat)
    x = sqrt(nextfloat(zero(T)))*2
    @test hypot(x,x/4) ≈ x * sqrt(17/T(16))
    @test hypot(x,x/4,x/4) ≈ x * sqrt(9/T(8))
end
```

thchr commented on base/math.jl:
```diff
@@ -536,38 +536,52 @@ Stacktrace:
+	N == 0 && error("invalid hypot() call without arguments")
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	r = sqrt(s)
```
Mustn't the type of `r = sqrt(s)` exactly match that of `s`, given that `s` is always a positive float (since `s = sum(abs2, float.(x))`). (similarly, if `s` is `Inf` or `NaN`, then `r = sqrt(s)` will also be)

cossio replied:
Ah your point is that `typeof(r) == typeof(s)`?

stevengj commented on test/math.jl:
```diff
@@ -1020,6 +1020,11 @@ using .Main.Furlongs
+        x = sqrt(nextfloat(zero(T)))*2
+        @test hypot(x,x/4) ≈ x * sqrt(17/BigFloat(16))
+        @test hypot(x,x/4,x/4) ≈ x * sqrt(9/BigFloat(8))
+    end
```
See the updated (more stringent) version of this test in the comments.

cossio replied:
I left both versions.

stevengj commented on base/math.jl:
```diff
@@ -582,6 +582,10 @@ function _hypot(x::NTuple{N,<:Number}) where {N}
 end
+# FIXME: revisit when #30387 is resolved
+_floatmin(x::AbstractFloat) = _floatmin(typeof(x))
+_floatmin(::Type{T}) where {T<:AbstractFloat} = nextfloat(zero(T)) / eps(T)
```
A slight optimization: `_floatmin(::Type{T}) where {T<:IEEEFloat} = floatmin(T)`.   This gives the same results, but is faster for `Float16`.

simonbyrne replied:
I'm also not sure how I feel about defining a `_floatmin` just for this function. It seems like we need a better way to figure out "is this thing a float-ey object, but not an `AbstractFloat`"?

thchr commented on base/math.jl:
```diff
@@ -536,38 +536,56 @@ Stacktrace:
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	if isnan(s)
+		return any(isinf, x) ? oftype(s, Inf) : sqrt(s) # IEEE 754
```
You can just return `s` rather than `sqrt(s)`.

cossio replied:
Of course! Thanks.

cossio commented:
Should I worry about the AppVeyor fail? The error message seems completely unrelated to this PR.

Anything else I can do to improve this?

stevengj commented on base/math.jl:
```diff
@@ -536,38 +536,56 @@ Stacktrace:
+function hypot(x::Number, y::Number)
+	#= Keep specialized two argument version, because it is faster
+	than generic varargs version below =#
+	ay, ax = minmax(abs(x), abs(y))
```
fix indenting here?

stevengj replied:
Also below… in general we indent by 4 spaces.

cossio replied:
Done. Thanks.

stevengj commented on base/math.jl:
```diff
@@ -536,38 +536,56 @@ Stacktrace:
+		end
+   	else
+   		return rr
+   	end
```
`oftype(rr, r)` seems better here, so that we don't have to construct a new `NaN` value.  (Not all `NaN` values are the same.)

cossio replied:
Done, thanks again!

stevengj commented on base/math.jl:
```diff
@@ -536,38 +536,56 @@ Stacktrace:
+function _hypot(x::NTuple{N,<:Number}) where {N}
+	N == 0 && error("invalid hypot() call without arguments")
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
```
`float.(x)` allocates a new tuple, no?  Seems better to do `sum(y -> abs2(float(y)), x)` so that there is a single loop and no temporaries.

cossio replied:
`Tuple`s are immutable, they are not allocated (on the heap). I tested the performance of these variants,

```julia
julia> fsum(x...) = sum(abs2, float.(x));
julia> hsum(x...) = sum(abs2.(float.(x)));
julia> gsum(x...) = sum(y -> abs2(float(y)), x);

julia> @btime fsum(1.,2.,3.,4.,5.,6.,7.,8.)
  8.116 ns (0 allocations: 0 bytes)
204.0
julia> @btime gsum(1.,2.,3.,4.,5.,6.,7.,8.)
  10.907 ns (0 allocations: 0 bytes)
204.0
julia> @btime hsum(1.,2.,3.,4.,5.,6.,7.,8.)
  2.436 ns (0 allocations: 0 bytes)
204.0
```

cossio replied:
It seems `hsum` is noticeably faster.

stevengj replied:
Weird, can you file a performance issue for this?  The `gsum` variant should not be so slow.

stevengj commented on base/math.jl:
```diff
@@ -536,38 +536,56 @@ Stacktrace:
+	# speculatively try fast naive approach
+	s = sum(abs2, float.(x))
+	if isnan(s)
+		return any(isinf, x) ? oftype(s, Inf) : s # IEEE 754
```
This is the wrong return type.  If `x` is dimensionful, then hypot should return something with the same units as `x`, but this returns a value with the units of `x^2`.

Just do `oftype(sqrt(s), Inf) : sqrt(s)` — the extra `sqrt` call is harmless since we don't need to optimize the NaN case, and the `oftype` branch will probably optimize away the call anyway.

cossio replied:
We wanted to avoid the extra `sqrt` call. In a previous commit we had `r = sqrt(s)` at the top, and then used `r` to decide the return type in all branches. I just benchmarked this and it is slower (seems the compiler is not able to recognize that we only need the type of `r`, not its value). I also benchmarked a version of the code where I used `sqrt(s)` instead of `r` wherever it was needed, but still the compiler is not optimizing the call away. Any suggestions?

cossio replied:
How can I write a test to with a dimensionful `x`?

stevengj replied:
For testing dimensionful `x`, see the tests with `Furlongs` in `tests/math.jl`.  There actually are a few tests for `hypot` already.

cossio replied:
Ah, that's what the "Furlong" was. It was only testing the two-arg version. I added some var args tests. Thanks

stevengj commented on base/math.jl:
```diff
@@ -536,38 +536,56 @@ Stacktrace:
+		if iszero(ma) || isinf(ma)
+			return oftype(s, ma)
+		else
+			return oftype(s, ma * sqrt(sum(abs2, x ./ ma)))
```
Again, `x ./ ma` allocates a temporary.   Better to do `sum(y -> abs2(y / ma), x)` so that there is a single loop.

To work around #15276, not sure if we need to wrap this in a `let` block.  i.e. `let ma = maximum(abs, x) ....`, to get the best performance?   (Nowadays #15276 doesn't seem to affect so many cases, but it would be good to check.)

cossio replied:
I think we can just use the dotted calls here. Tuples do not get allocated. Also `@btime` suggests that this way is faster.

stevengj replied:
Or, pursuant to your comment above, `sum(abs2.(x ./ ma))`, if there is a performance bug with the `sum(y -> ..., ...)` version in the current Julia.

cossio replied:
Done.

stevengj commented on base/math.jl:
```diff
@@ -566,7 +566,7 @@ hypot(x::Number...) = _hypot(promote(x...))
 	N == 0 && error("invalid hypot() call without arguments")
 	# speculatively try fast naive approach
-	s = sum(abs2, float.(x))
+	s = sum(abs2.(float.(x)))
```
I'm confused that this would be faster than `sum(y -> abs2(float(y)), x)`, which avoids creating a temporary tuple.

cossio replied:
I think the compiler avoids creating a new tuple and probably does the thing in-place. If it `x` were an `Array`, it would be a different story.

stevengj commented on base/math.jl:
```diff
@@ -536,38 +536,57 @@ Stacktrace:
+        return any(isinf, x) ? oftype(r, Inf) : s # IEEE 754
+    elseif isinf(s) || s ≤ _floatmin(s) * (N-1) # FIXME: revisit when #30387 is resolved
+        # if overflow/underflow, try normalization
+        ma = maximum(abs, x)
```
I think you need `float(maximum(abs, x))` here.

* Then you can eliminate the `oftype` calls, so that you no longer need `r` to be computed (i.e. you eliminate the `sqrt` call).

* It looks like your current code can overflow for rationals, since `x ./ ma` will produce another rational that can overflow in the `abs2`.  In particular, consider `let x = typemax(Int); @test hypot(x//1, x//1, x//1) ≈ x*√3; end`

cossio replied:
Thanks! I added this test case

cossio commented on base/math.jl:
```diff
@@ -536,38 +536,57 @@ Stacktrace:
+    N == 0 && error("invalid hypot() call without arguments")
+    # speculatively try fast naive approach
+    s = sum(abs2.(float.(x)))
+    r = sqrt(s)
```
You are suggesting this?

```julia
function _hypot(x::NTuple{N,<:Number}) where {N}
    N == 0 && error("invalid hypot() call without arguments")
    s = sum(abs2.(float.(x)))
    if isnan(s)
        return any(isinf, x) ? oftype(sqrt(s), Inf) : sqrt(s) # IEEE 754
    elseif isinf(s) || s ≤ _floatmin(s) * (N-1)
        ma = maximum(abs, x)
        if iszero(ma) || isinf(ma)
            return oftype(sqrt(s), ma)
        else
            return oftype(sqrt(s), ma * sqrt(sum(abs2.(x ./ ma))))
        end
    else
        return sqrt(s)
    end
end
```

`sqrt(s)` is called in every branch, right? Or am I missing something? 
Also note that I think that `oftype(sqrt(s), ...)` doesn't optimize the call away.

cossio replied:
Nevermind the previous comment. See new changes.

stevengj commented:
Actually, I think the `maximum(abs, x)` computation is wrong for both the 2-arg and 3-arg versions.  We need `maximum(y -> abs(float(x)), x)` or equivalent to handle negative integers properly.

In particular, thanks to 2s complement, we have:
```jl
julia> typemin(x)
-9223372036854775808

julia> abs(typemin(x))
-9223372036854775808
```

As a result, even our two-argument `hypot` function is broken:
```jl
julia> hypot(typemin(Int), typemin(Int))
-1.3043817825332783e19
```

cossio commented on base/math.jl:
```diff
@@ -536,38 +536,57 @@ Stacktrace:
+    s = sum(abs2.(float.(x)))
+    r = sqrt(s)
+    if isnan(s)
+        return any(isinf, x) ? oftype(r, Inf) : s # IEEE 754
```
Corrected. This should also be catched now by the `Furlong` varargs tests.

cossio commented on test/math.jl:
```diff
@@ -985,3 +984,53 @@ using .Main.Furlongs
+@test hypot(Furlong(0), Furlong(0), Furlong(0)) == Furlong(0.0)
+@test hypot(Furlong(1), Furlong(1), Furlong(1)) == Furlong(sqrt(3))
+@test hypot(Furlong(Inf), Furlong(NaN), Furlong(0)) == Furlong(Inf)
+@test hypot(Furlong(Inf), Furlong(Inf), Furlong(Inf)) == Furlong(Inf)
```
Done.

cossio commented:
You're right. `abs(typemin(Int))` fails because the maximum integer is one unit smaller than the minimum.
We need to convert to `float` first.

I added new commits and a test with this.

> Alternatively I guess abs(float(maximum(abs, x))) might be faster.

This has the same issue, right?

cossio commented:
By the way, the number of tests has diverged. I don't know if it makes sense to keep them all or if we should try to clean them up a bit.

stevengj commented:
`abs(float(maximum(abs, x)))` does not have the same issue.  `abs(float(abs(typemin(Int)))) == abs(float(typemin(Int)))`

cossio commented:
@stevengj I mean that

```julia
julia> abs(float(maximum(abs, (typemin(Int), 1))))
1.0
julia> maximum(abs∘float, (typemin(Int), 1))
9.223372036854776e18
```

stevengj commented on test/math.jl:
```diff
@@ -980,8 +979,62 @@ end
+        @test @inferred hypot(i, i, i, i) ≈ 2.0i
+        @test @inferred hypot(i//1, 1//i, 1//i) ≈ i
+    end
+    @test @inferred hypot(typemin(Int), typemin(Int)) ≈ -sqrt(2)typemin(Int)
```
Also need a test for the 3-arg `typemin` case.

stevengj commented:
Thanks, looking good.  I wouldn't worry about having too many tests.

stevengj commented on test/math.jl:
```diff
@@ -1008,7 +1008,11 @@ using .Main.Furlongs
+    let i = typemin(Int)
+        @test @inferred hypot(i, i) ≈ -√2i
+        @test @inferred hypot(i, i, i) ≈ -√3i
+        @test @inferred hypot(i, i, i, i) ≈ -2.0i
```
These are the wrong answers — the result can't be negative.

cossio replied:
It is not. Note that `typemin(Int)` is negative.

stevengj commented:
Is it allocating for more than 4 arguments, as in https://github.com/JuliaLang/julia/issues/30465#issuecomment-449124730?

cossio commented:
Doesn't seem like it:

```julia
julia> @btime hypot(1.,2.,3.,4.,5.,6.,7.,8.,9.,10.)
  23.943 ns (0 allocations: 0 bytes)
19.621416870348583
```

Though it allocates if the tuple has heterogeneous types:

```julia
julia> @btime hypot(1,2,3,4.,5.,6.,7.,8.,9.,10.)
  316.980 ns (16 allocations: 976 bytes)
19.621416870348583
```

But this (the heterogeneous tuple case) also allocated in every other `hypot` version I've tried (with/without dotted calls, etc... and the original version before this PR).

cossio commented:
The `Furlong` tests are failing and I think the cause is that `float(Furlong(1)) == 1.0` strips the unit. I'm not sure how this should go.

Is this a fault of `Furlong` (it should define a different method for `float`)?

Or is this something I should fix here?

stevengj commented:
I think we need to define `float(x::Furlong) = Furlong(float(x))` in `Furlongs.jl`

stevengj commented on base/math.jl:
```diff
@@ -536,38 +536,56 @@ Stacktrace:
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = _hypot(promote(x...))
+function _hypot(x::NTuple{N,<:Number}) where {N}
```
Now that #30471 is merged, maybe we should try getting rid of the `promote`.  That is, just define
```
function hypot(x::Number...)
   isempty(x) && error("invalid hypot() call without arguments")
   s = sum(x -> abs2(float(x)), x)
    if isnan(s)
        return any(isinf, x) ? oftype(sqrt(s), Inf) : sqrt(s) # IEEE 754
    elseif isinf(s) || s ≤ _floatmin(s) * (N-1) # FIXME: revisit when #30387 is resolved
        # if overflow/underflow, try normalization
        ma = maximum(x -> abs(float(x)), x)
       ...
```

cossio replied:
Thanks!  I think that does it. Let's see what CI reports.

StefanKarpinski commented on test/math.jl:
```diff
@@ -1008,7 +1008,11 @@ using .Main.Furlongs
     end
-    @test @inferred hypot(typemin(Int), typemin(Int)) ≈ -sqrt(2)typemin(Int)
+    let i = typemin(Int)
+        @test @inferred hypot(i, i) ≈ -√2i
```
I think you want `@test @inferred(hypot(i, i)) ≈ -√2i` here since presumably you want to test specifically that `hypot` is inferred, not the entire expression `hypot(i, i) ≈ -√2i`.

cossio commented on base/math.jl:
```diff
@@ -536,38 +536,54 @@ Stacktrace:
     else
         r = ay / ax
     end
+    rr = ax * sqrt(oneunit(r) + r * r)
```
Then I can just use `one(r)`... I'm not sure why I put `oneunit` here ...

cossio commented on base/math.jl:
```diff
@@ -536,38 +536,54 @@ Stacktrace:
+function hypot(x::Number...)
+    N == 0 && error("invalid hypot() call without arguments")
+    # speculatively try fast naive approach
+    s = sum(abs2.(float.(x)))
```
I did that change already. Thanks!

stevengj replied:
It doesn't look like you did that change already — `sum(abs2.(float.(x)))` still shows up in the diff.

thchr commented on base/math.jl:
```diff
@@ -562,13 +562,13 @@ end
 """
-function hypot(x::Number...) where {N}
+function hypot(x::Number...)
     N == 0 && error("invalid hypot() call without arguments")
```
No, I mean the `N` in `N == 0 && error(...)` in the line above should be swapped for `length(x)`.

stevengj commented:
Have you verified that the performance is still good?  Hopefully better now for mixed types, too.

cossio commented:
> Have you verified that the performance is still good? Hopefully better now for mixed types, too.

No, I'm not sure how to do it. I've been testing this on the stable julia 1.0.3 which doesn't include your PR.
How can I test this on the latest julia master? Ideally without having to recompile julia after every change I make.

stevengj commented:
You can just `import Base: mapfoldl_impl` and paste [my code from tuple.jl](https://github.com/stevengj/julia/blob/0b46d825be9a6069d80403c81ce1b7967697b986/base/tuple.jl#L215-L226) into the REPL, I think?

cossio commented:
> I think we need to define `Base.float(x::Furlong{p}) where {p} = Furlong{p}(float(x.val))` in `Furlongs.jl`

Maybe this is not the best approach. Because then I have no way of stripping the unit.
Instead I'm using `_floatunit(x) = float(x) * oneunit(x)`. Let me know what you think.

cossio commented:
Since `hypot(Inf, NaN)` is `Inf`, probably `hypot(missing, Inf)` should also be `Inf`, right?

stevengj commented:
I think the `missing` case should be a separate PR.  

`float(::Furlong)` should certainly return a `Furlong`.  The `float` function does *not* return an `AbstractFloat` in general, as can be seen from its treatment of arrays, complex numbers, etcetera, and is also what Unitful.jl does.

So you should use `s = sum(abs2 ∘ float, x)` and get rid of `_floatunit`.   You can then do `_floatmin(one(s))`, since `one` strips out the dimensions.

stevengj commented on base/math.jl:
```diff
@@ -564,10 +564,10 @@ Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
-    fs = float(s)
-    if isinf(s) || fs ≤ _floatmin(fs) * (length(x) - 1)
+    sf = AbstractFloat(s)
+    if isinf(s) || sf ≤ _floatmin(sf) * (length(x) - 1)
```
`AbstractFloat(s)` won't necessarily exist.  If you want to fix the units in the comparison, you can do `s ≤ oneunit(s) * _floatmin(sf) * (length(x) - 1)`

cossio replied:
Done.

cossio commented:
@stevengj @thchr Sorry for dropping the ball on this for some time. Care to take one last look?

There is also https://github.com/JuliaLang/julia/pull/31922, but I think that doesn't conflict with this, since it only deals with the two-argument case and adds an extra method for specific types of numbers.

stevengj commented on base/math.jl:
```diff
@@ -537,38 +537,61 @@ Stacktrace:
+hypot() = error("invalid hypot() call without arguments")
+hypot(x::Number) = abs(float(x))
+
+_hypot_squared(x::Tuple{Vararg{Number}}) = sum(abs2 ∘ float, x)
```
Where is this function used?

stevengj replied:
Oh, I see it above

simonbyrne replied:
since it is only used once, why not just inline it?

cossio commented:
@stevengj @simonbyrne @StefanKarpinski ready to merge?

simonbyrne commented on base/math.jl:
```diff
@@ -537,38 +537,61 @@ Stacktrace:
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+function hypot(x::Number...)
```
It may be possible to avoid the need for `hypot()` by defining this as `hypot(x::Number, xs::Number...)`. 

Also, `x` isn't a good variable name for a tuple (since it is used above as the variable name of a scalar.

simonbyrne commented on base/math.jl:
```diff
@@ -537,38 +537,61 @@ Stacktrace:
+    end
+end
+
+hypot() = error("invalid hypot() call without arguments")
```
If we can't avoid defining this, it should throw a `MethodError`

simonbyrne commented on base/math.jl:
```diff
@@ -537,38 +537,61 @@ Stacktrace:
+hypot(x::Number) = abs(float(x))
+
+_hypot_squared(x::Tuple{Vararg{Number}}) = sum(abs2 ∘ float, x)
+function _hypot_safe(x)
```
same, this can be inlined

thchr commented:
What did the final benchmark timings end up like?

cossio commented:
I fixed the comments @simonbyrne @stevengj .
Using the example benchmark we were using at the beginning I get @thchr 

```
julia> @btime hypot(3,4,5)
  1.539 ns (0 allocations: 0 bytes)
7.0710678118654755
```

which is at least as fast as the original implementation.
If there are no further issues (and tests pass) can we get this merged? The original implementation is broken anyway so this is a net improvement.

cossio commented:
It looks like there have been some nice compiler improvements!
The following simpler code:

```julia
hypot(x::Number, xs::Number...) = _hypot(float.(promote(x, xs...))...)
function _hypot(x::T...) where {T<:Number}
    maxabs = maximum(abs, x)
    isnan(maxabs) && any(isinf, x) && return T(Inf)
    (iszero(maxabs) || isinf(maxabs)) && return maxabs
    return maxabs * sqrt(sum(y -> abs2(y / maxabs), x))
end
```

Passes all tests and is as fast or faster than the current implementation (even covering the two-args version). What do you guys think?

cossio commented:
Turns out the doctest failing was important. `hypot(3, 4im)` should return real 5, but was returning complex `5 + 0im`. I fixed that and added two examples like this as ordinary tests.

Commit:
simplify hypot

cossio commented:
I went ahead a committed the simplified varargs version, keeping the more complicated version with a separate two-args method in a separate commit.

I think we should just go ahead with this simpler code.

cossio commented:
CI is failing with an ambiguity error that I don't understand. Someone care to take a look? @StefanKarpinski ?

cfborges commented:
I would make two general comments:

First: The current two argument hypot code achieves (or nearly achieves if there is no fma) a level of accuracy that is requested in the IEEE754 standard (i.e. correctly rounded result). There is no reason to undo that and replace it with an algorithm that doesn't.

Second: The IEEE754 standard indicates that hypot should be a two argument function. I am of the opinion that more than two arguments leads to the 2-norm and that should be a different function. 

One man's point of view. Your mileage may vary.

cossio commented:
If there is a test-case where the current two-args version achieves more accuracy, it would be nice to add it to the tests explicitly. Can you suggest an example @cfborges? I was actually surprised the simple implementation passed all the tests that are already there.

As to the multiple args version, I agree that we can just use LinearAlgebra.norm. But for some reason the multi-args hypot was present before this PR and I think removing it at this point would be a breaking change.

Commit:
combine with correctly rounded 2 arg version

Keno commented:
This needs a rebase, but @stevengj I assume you're still in favor?
